Assignment2 Vibrato

talker93 opened the pull request:
Hi all, thanks for your reviewing!

alexanderlerch committed:
ringbuffer class with stub functions to fill in

alexanderlerch committed:
removed accidentally committed .bak files and updated gitignore

alexanderlerch committed:
Ringbuffer solution

alexanderlerch committed:
Merge branch 'exercise_ringbuff'

# Conflicts:
#	src/MUSI6106Exec/MUSI6106Exec.cpp

alexanderlerch committed:
assert arguments where incorrect in AudioFile.cpp, fixes #5

alexanderlerch committed:
added tests and gtest testing framework with continuous integration

added Vibrato project and empty files

alexanderlerch committed:
added doxygen and updated README.md

alexanderlerch committed:
named test namespaces

alexanderlerch committed:
fixed branch name

Jas1238Smith committed:
Remove "static" from audiofile_test members.

talker93 committed:
first dev

talker93 committed:
Merge branch 'assignment2_vibrato' of https://github.com/alexanderlerch/2022-MUSI6106 into assignment2_vibrato

jocekav committed:
Update .gitignore

jocekav committed:
Add vibrato test framework and test cases

talker93 committed:
finally running

jocekav committed:
Add initial RingBuffer tests and fix syntax errors

jocekav committed:
Merge branch 'as2_vibrato' of https://github.com/talker93/2022-MUSI6106 into as2_vibrato

jocekav committed:
Additional ring buffer tests

jocekav committed:
Converted vibrato tests to gtests

jocekav committed:
Update repo to remain consistent with Lerch's current repo

jocekav committed:
Converted ring buffer to gtests and successful build for testing target

talker93 committed:
Added Matlab results

talker93 committed:
Create vibrato2.jpg

talker93 committed:
Update README.md

talker93 committed:
Update README.md

jocekav committed:
Fixed ModAmp0 and DCEqual tests

jocekav committed:
Merge branch 'as2_vibrato' of https://github.com/talker93/2022-MUSI6106 into as2_vibrato

jocekav committed:
Add additional vibrato test

talker93 committed:
Rewrote the vibrato.h and vibrato.cpp

13 tests passed.
2 tests failed.

talker93 committed:
Solved the inputValid issue in the test

likelian commented on src/inc/Lfo.h:
```diff
@@ -0,0 +1,88 @@
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        } else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
```
Is m_fModFreq a constant that will not change once the instance is created?

talker93 replied:
No it will be updated when you call out the CLfo::setParam function.
But, besides that, I just found I lost a lot changes which hadn't been pushed.
In my newest version, it actually execute an reset() and init() evrery time when you want to use CVibrato::setParam function.

talker93 replied:
The lost setParam/getParam function added, check the newest version.

likelian commented on src/inc/Lfo.h:
```diff
@@ -0,0 +1,88 @@
+    {
+        float *pSine = new float [m_iBufferLength];
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
```
Is the output computed by sin() function or by different read speed of wavetable?

talker93 replied:
pSine only stores sinewave value in one cycle, but its length would affect the reading speed, as the read index increase one place when input buffer read index increase one place.

talker93 replied:
Now that in the newest version, we get sinewave value in a more clear way.
```
Lfo.h - Line to 60
    Error_t generate () {
        CSynthesis::generateSine(m_pSine, m_fModFreqRatio * m_fSampleRate, m_fSampleRate, m_iWaveLength);
        m_pRingBuffer -> putPostInc(m_pSine, m_iWaveLength);
    }
    
    float getValue () {
        return m_pRingBuffer -> getPostInc();
    }
```

JohnKasun commented on src/inc/Lfo.h:
```diff
@@ -0,0 +1,88 @@
+class CLfo
+{
+public:
+    CLfo(float fModFreq, float fSampleRate)
```
I think the inclusion of an amplitude parameter in this constructor would be a good addition. That way you can handle the LFO's amplitude within this class rather than externally

talker93 replied:
I'm not sure if I made this correctly.
The amplitude in my program is controlled by the mod width, which is one of the input in the main function.
```
Vibrato.cpp - line 118
Error_t CVibrato::process(float **ppfInputBuffer, float **ppfOutputBuffer, int iNumberOfFrames) {
    float fModAmp, fDelay = 0;
    for(int i = 0; i < m_iNumChannels; i++) {
        for(int j = 0; j < iNumberOfFrames; j++) {
            fModAmp = m_pCLfo -> getValue();
            fDelay = 1 + m_fDelayInS * m_fSampleRateInHz + m_fModWidthInS * m_fSampleRateInHz * fModAmp;
            m_ppCRingBuffer[i] -> putPostInc(ppfInputBuffer[i][j]);
            ppfOutputBuffer[i][j] = m_ppCRingBuffer[i] -> get(fDelay);
            m_ppCRingBuffer[i] -> getPostInc();
        }
    }
    return Error_t::kNoError;
}
```

JohnKasun commented on src/inc/Lfo.h:
```diff
@@ -0,0 +1,88 @@
+        m_pCRingBuffer = 0;
+    }
+    
+    Error_t setParam(float fParamValue)
+    {
+        m_fModFreq = fParamValue;
+        float *pSine = new float [m_iBufferLength];
+        
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        }
+        else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
+        m_pCRingBuffer-> putPostInc(pSine, m_iBufferLength);
+        delete [] pSine;
+        
+        return Error_t::kNoError;
+    }
```
It looks like the only settable parameter here is the mod frequency. Changing the function definition to maybe `Error_t setParam(Lfo::LfoParam_t eParam, float fParamValue)` and then having the option to change parameters like the sample rate and amplitude would be more useful I think

talker93 replied:
Since LFO is not a big project, I didn't apply the setParam function, instead, you can reset() then init() with your parameters then generate(). That way you should be able to set values.
```
Lfo.h - Line 33
    Error_t init (float fModFreqRatio, float fSampleRate) {
        this -> reset();
        if (fModFreqRatio <= 0 || fSampleRate <= 0) {
            return Error_t::kFunctionInvalidArgsError;
        }
        m_bIsInitialized = true;
        m_fModFreqRatio = fModFreqRatio;
        m_fSampleRate = fSampleRate;
        m_iWaveLength = static_cast<int>(1/fModFreqRatio);
        m_pSine = new float [m_iWaveLength];
        m_pRingBuffer = new CRingBuffer<float>(m_iWaveLength);
        return Error_t::kNoError;
    }
    
    Error_t reset () {
        if (m_bIsInitialized == true) {
            delete m_pRingBuffer;
            m_pRingBuffer = 0;
            delete [] m_pSine;
            m_pSine = 0;
        }
        m_fSampleRate = 0;
        m_fModFreqRatio = 0;
        m_bIsInitialized = false;
        return Error_t::kNoError;
    }
    
    Error_t generate () {
        CSynthesis::generateSine(m_pSine, m_fModFreqRatio * m_fSampleRate, m_fSampleRate, m_iWaveLength);
        m_pRingBuffer -> putPostInc(m_pSine, m_iWaveLength);
    }
```

likelian reviewed (commented):
Good job!

JohnKasun commented on src/inc/Lfo.h:
```diff
@@ -0,0 +1,88 @@
+        
+    }
+    
+    void process()
+    {
+        float *pSine = new float [m_iBufferLength];
+        
+        CSynthesis::generateSine(pSine, m_fModFreq * m_fSampleRate, m_fSampleRate, m_iBufferLength);
+        
+        m_pCRingBuffer-> putPostInc(pSine, m_iBufferLength);
+        
+        delete [] pSine;
+    }
```
Having memory allocation and deletion and generating a sinusoid "m_iBufferLength" long each time the process called may not be the most optimal. Also, you could definitely condense your `float returnLfoVal()` function and your `void process()` function into a single `float process()` function

talker93 replied:
Thanks for advising, now it's improved!
```
Vibrato.h - Line 118
Error_t CVibrato::process(float **ppfInputBuffer, float **ppfOutputBuffer, int iNumberOfFrames) {
    float fModAmp, fDelay = 0;
    for(int i = 0; i < m_iNumChannels; i++) {
        for(int j = 0; j < iNumberOfFrames; j++) {
            fModAmp = m_pCLfo -> getValue();
            fDelay = 1 + m_fDelayInS * m_fSampleRateInHz + m_fModWidthInS * m_fSampleRateInHz * fModAmp;
            m_ppCRingBuffer[i] -> putPostInc(ppfInputBuffer[i][j]);
            ppfOutputBuffer[i][j] = m_ppCRingBuffer[i] -> get(fDelay);
            m_ppCRingBuffer[i] -> getPostInc();
        }
    }
    return Error_t::kNoError;
}
```
```
Lfo - Line 65
    float getValue () {
        return m_pRingBuffer -> getPostInc();
    }
```

JohnKasun commented on src/inc/Lfo.h:
```diff
@@ -0,0 +1,88 @@
+    //return current values
+    float returnLfoVal()
+    {
+        float fLfo = m_pCRingBuffer->get(m_fRead);
+        
+        m_fRead = m_fRead + 1;
+        
+        if (m_fRead >= m_iBufferLength) {
+            m_fRead -= m_iBufferLength;
+        }
+        
+        return fLfo;
```
Not sure if this was intentional or not, but there doesn't appear to be any fractional indexing occuring with `m_pCRingBuffer->get(m_fRead)` since `m_fRead` will always be an integer

talker93 replied:
```
Error_t CVibrato::process(float **ppfInputBuffer, float **ppfOutputBuffer, int iNumberOfFrames) {
    float fModAmp, fDelay = 0;
    for(int i = 0; i < m_iNumChannels; i++) {
        for(int j = 0; j < iNumberOfFrames; j++) {
            fModAmp = m_pCLfo -> returnLfoVal();
            fDelay = 1 + m_fDelayInS * m_fSampleRateInHz + m_fModWidthInS * m_fSampleRateInHz * fModAmp;
            m_ppCRingBuffer[i] -> putPostInc(ppfInputBuffer[i][j]);
            ppfOutputBuffer[i][j] = m_ppCRingBuffer[i] -> get(fDelay);
            m_ppCRingBuffer[i] -> getPostInc();
        }
    }
    
    return Error_t::kNoError;
    
}
````
The returned value from CLfo is merely a sinewave data serving as a modulation amplititude. The fraction delay is used in CVibrato::process function.

talker93 replied:
These lines of code are simply getting value from the sinewave, while the fractional index is being used at in CVibrato::process().
```
Vibrato.cpp - Line 125
ppfOutputBuffer[i][j] = m_ppCRingBuffer[i] -> get(fDelay);
```

JohnKasun commented on src/inc/Lfo.h:
```diff
@@ -0,0 +1,88 @@
+{
+public:
+    CLfo(float fModFreq, float fSampleRate)
+    {
+        m_pCRingBuffer = 0;
+        m_fRead = 0;
+        m_fModFreq = fModFreq;
+        m_fSampleRate = fSampleRate;
+        if (m_fModFreq == 0) {
+            m_iBufferLength = 1;
+        } else {
+            m_iBufferLength = static_cast<int>(1 / m_fModFreq);
+        }
+        m_pCRingBuffer = new CRingBuffer<float>(m_iBufferLength);
+    }
```
Within this constructor is where I would have generated all the values for the wavetable and put them in `m_pCRingBuffer`. Then, have `currentIndex` and `indexIncrement` variables that represent your current position in and current frequency of the wavetable. This will allow you to simply call `m_fCRingBuffer.get(currentIndex)` within your `process()` to get the appropriate value

talker93 replied:
I just renewed the function, now it looks pretty concise.
```
Lfo.h - Line 65
    Error_t generate () {
        CSynthesis::generateSine(m_pSine, m_fModFreqRatio * m_fSampleRate, m_fSampleRate, m_iWaveLength);
        m_pRingBuffer -> putPostInc(m_pSine, m_iWaveLength);
    }
    
    float getValue () {
        return m_pRingBuffer -> getPostInc();
    }
```

JohnKasun commented on src/Vibrato/Vibrato.cpp:
```diff
@@ -0,0 +1,108 @@
+    return Error_t::kNoError;
+}
+
+Error_t CVibrato::process(float **ppfInputBuffer, float **ppfOutputBuffer, int iNumberOfFrames) {    float fModAmp, fDelay = 0;
```
I think also passing in an `iNumChannels` parameter would be good so you don't reference unallocated values

ijc8 replied:
I guess this is assuming the user will pass in buffers that concur with their previous `init()` call, since it wouldn't make sense to call `process` twice with different numbers of channels (without resetting/re-initializing the vibrato in between).

talker93 replied:
Yes, all other parameters like the basic delay, mod freq and sample rate are also used in process() function. I think it's safer to get all those values in either init() or setParam();

ijc8 commented on src/MUSI6106Exec/MUSI6106Exec.cpp:
```diff
@@ -1,139 +1,132 @@
+    // arguments should be in this sequence: (input_file_path output_file_path sample_rate mod_frequency mod_delay mod_width)
+    
+    if(argc < 6) {
+        cout << "missing some arguments" << endl;
```
This error message is accurate, but it's also not very helpful.
You could consolidate this with the comment above by printing something like "usage: <input_file_path> <output_file_path> ...". This would allow you to remove the comment and help the user out by reminding them of the expected arguments.

talker93 replied:
Thanks, I didn't carefully check the comparing content, I will probably create new branch or update the main branch for comparing for next time.

talker93 replied:
```suggestion
        cout << "missing some arguments" << endl;
        cout << "Usage: <input_path> <sample_rate> <mod_frequency> <delay> <mod_width>" << endl;
        cout << "Example: ./sweep.wav 48000 1000 0.001 0.001" << endl;
```

The input hint now has been printed.

ijc8 commented on src/Tests/Tests/Test_Vibrato.cpp:
```diff
@@ -0,0 +1,296 @@
+            ppfTempIn[i]  = new float[iBlockSize];
+            ppfTempOut[i] = new float[iBlockSize];
+
+            CSynthesis::generateSine(ppfAudioData[i], fTestSigFreq, fSampleRate, iBlockSize);
```
Nice use of `SetUp()`, `TearDown()`, and library functions to avoid duplication!

talker93 replied:
Thanks!

ijc8 commented on src/Tests/Tests/Test_Vibrato.cpp:
```diff
@@ -0,0 +1,296 @@
+    pVibrato -> CVibrato::init(fDelay, fSampleRate, fModWidth, fModFreq, iNumChannels);
+    
+    testProcess(ppfOutputData, iBlockSize);
+    testProcess(ppfOutBlock, 459);
```
May be worth testing more than a single funky block size here.

ijc8 commented on src/Tests/Tests/Test_Vibrato.cpp:
```diff
@@ -0,0 +1,296 @@
+    
+    for (int i = 0; i < iNumChannels; i++)
+    {
+        CHECK_ARRAY_CLOSE(ppfOutBlock[i], ppfOutputData[i], iBlockSize, 1e-3F);
```
I'm a little confused here - `ppfOutBlock` is the same size as `ppfOutputData`, but only the first 459 frames should be filled with meaningful data (because of the `testProcess` call above). So why would `ppfOutBlock[i]` be close to `ppfOutputData[i]` when `i >= 459`?
Also - should the vibrato be reset between the two `testProcess` calls above? Otherwise it seems that the delay line and LFO will not start in the same state in those two cases.

ijc8 commented on src/inc/Vibrato.h:
```diff
@@ -0,0 +1,82 @@
+     \param eParam
+     \return Error_t
+     */
+    Error_t setParam (VibratoParam_t eParam) const;
```
I see that `setParam` is defined for `CLfo`, but I didn't see the implementation for `CVibrato` anywhere. Is it missing, or did I just miss it? Also, is there a corresponding `getParam`?

talker93 replied:
I re-wrote the code for multiple versions, and I just found some part of it I got lost in my repo.
As the setParam doesn't use in normal main process, it wasn't detected.

talker93 replied:
I will push a new version with setParam function. But the CLfo::setParam is not being used anymore.

talker93 replied:
Updated.

ijc8 reviewed (changes requested):
It would be nice if you updated your fork's `main` branch, or otherwise created this PR in reference to the starter code. As of now, this PR appears to contain 273 files changed, and 113,107 lines added, because it includes Alexander's commits leading up to the assignment.
(This has nothing to do with your solution, but it makes the code harder to review because reviewers have to get past all of the irrelevant changes before they can review what you did.)

Also, despite the .gitignore, it appears that some .DS_Stores still managed to sneak in from someone's Mac.

talker93 committed:
Major change

1. Added the lost part of setParam/getParam/etc function in Vibrato.cpp;
2. Re-write the CLfo class in general style, create->init->generate->getValue;
3. Refactor some misleading arguments name, such as the fModFreq in Lfo.h;

talker93 committed:
Update .gitignore

Add wav format as ignored.

jocekav committed:
Update tests based on peer reviews
